### ISSUE TITLE
[Feature] Node steps to run after SSH is available.

### ISSUE THREAD

**Issue by mitchellmaler** [rancher/rancher#13720]
Sometimes a VM image that is created by docker machine needs a few things ran on it to get it setup correctly or be ready kubernetes. Right now the choices are to pre-build the image with the custom setup or manually deploy with RKE after the VM is manually provisioned. It would be nice to have a custom script that can run right after the is available to run custom setup.

**Comment by alexandrem** [comment]
I thought about this, but need to dig a bit more.

The docker engine install script is not the correct place, because if docker is already present on the system (e.g CoreOs or some custom cloud images with prebuilt packages) then docker-machine will skip the execution of the script entirely (no `curl | sh`)

Ideally we need a new system setting to point to an external shell script, e.g `system-init-url`

I guess we'll have to support Windows at some point, so perhaps having a different setting per OS type.

Finally, I suppose this should go in the machine driver code, once the node is provisioned but before kontainer-engine rpc starts?

**Comment by mitchellmaler** [comment]
I agree. This seems like a feature that machine needs almost like a cloud-init type of script that runs after installing docker that can either be sh or ps1 depending on the OS type.

**Comment by cloudnautique** [comment]
This is pretty much already supported. Assuming you are using a Rancher node provisioner, in the node template under engine options there is a Docker install URL which is a URL to a script. The default is essentially the get.docker.com install script pinned to specific versions by us. But you can host your own and do whatever you want in it as long as in the end Docker is up and running. We have used this in the past internally to setup device mapper with lvm-thinpools for instance.

**Comment by alexandrem** [comment]
@cloudnautique

Like mentioned above the engine install script is not an option.

This cannot work for coreos.
